presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantTopNColumns.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.matching.Captures;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.spi.plan.LogicalProperties;
+import com.facebook.presto.spi.plan.OrderingScheme;
+import com.facebook.presto.spi.plan.TopNNode;
+import com.facebook.presto.sql.planner.iterative.GroupReference;
+import com.facebook.presto.sql.planner.iterative.Rule;
+
+import static com.facebook.presto.sql.planner.iterative.rule.RemoveRedundantSortColumns.pruneOrderingColumns;
+import static com.facebook.presto.sql.planner.plan.Patterns.topN;
+
+/**
+ * Removes TopN columns from input if the source has a Key that refers to the ordering columns
+ */
+public class RemoveRedundantTopNColumns
+        implements Rule<TopNNode>
+{
+    private static final Pattern<TopNNode> PATTERN = topN().matching(p -> ((GroupReference) p.getSource()).getLogicalProperties().isPresent());
+
+    @Override
+    public Pattern<TopNNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(TopNNode node, Captures captures, Context context)
+    {
+        OrderingScheme orderingScheme = node.getOrderingScheme();
+
+        LogicalProperties sourceLogicalProperties = ((GroupReference) node.getSource()).getLogicalProperties().get();
+        OrderingScheme newOrderingScheme = pruneOrderingColumns(orderingScheme, sourceLogicalProperties);
```

Comment:
This is reusing a method defined in RemoveRedundantSortColumns, can you build a common class and have these two to inherit, or simply move this method to a util class?

Reply:
>Why is it a TODO but not part of this PR?  

I did add a presto-tests test. Forgot to update the PR description

Reply:
Chose to move the static methods to a Util class